sw/device/lib/crypto/drivers/entropy.c

```diff
@@ -0,0 +1,183 @@
+  reg = bitfield_field32_write(0, kAppCmdFieldCmdId, cmd.id);
+  reg = bitfield_field32_write(reg, kAppCmdFieldCmdLen, cmd_len);
+  reg = bitfield_field32_write(reg, kAppCmdFieldGlen, cmd.generate_len);
+
+  if (launder32(cmd.disable_trng_input) == kHardenedBoolTrue) {
+    reg = bitfield_bit32_write(reg, kAppCmdBitFlag0, true);
+  }
+
+  abs_mmio_write32(reg_address, reg);
+
+  for (size_t i = 0; i < cmd_len; ++i) {
+    abs_mmio_write32(reg_address, cmd.seed_material->data[i]);
+  }
+  return OK_STATUS();
+}
+
+status_t entropy_csrng_instantiate(
+    hardened_bool_t disable_trng_input,
+    const entropy_seed_material_t *seed_material) {
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpInstantiate,
+                                .disable_trng_input = disable_trng_input,
+                                .seed_material = seed_material,
+                                .generate_len = 0,
+                            });
+}
+
+status_t entropy_csrng_reseed(hardened_bool_t disable_trng_input,
+                              const entropy_seed_material_t *seed_material) {
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpReseed,
+                                .disable_trng_input = disable_trng_input,
+                                .seed_material = seed_material,
+                                .generate_len = 0,
+                            });
+}
+
+status_t entropy_csrng_update(const entropy_seed_material_t *seed_material) {
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpUpdate,
+                                .seed_material = seed_material,
+                                .generate_len = 0,
+                            });
+}
+
+status_t entropy_csrng_generate_start(
+    const entropy_seed_material_t *seed_material, size_t len) {
+  // Round up the number of 128bit blocks. Aligning with respect to uint32_t.
+  // TODO(#6112): Consider using a canonical reference for alignment operations.
+  const uint32_t num_128bit_blocks = (len + 3) / 4;
+  return csrng_send_app_cmd(kBaseCsrng + CSRNG_CMD_REQ_REG_OFFSET,
+                            (entropy_csrng_cmd_t){
+                                .id = kEntropyDrbgOpGenerate,
+                                .seed_material = seed_material,
+                                .generate_len = num_128bit_blocks,
+                            });
+}
+
+status_t entropy_csrng_generate_data_get(uint32_t *buf, size_t len) {
+  for (size_t i = 0; i < len; ++i) {
+    // Block until there is more data available in the genbits buffer. CSRNG
+    // generates data in 128bit chunks.
+    if (i % kEntropyCsrngBitsBufferSize == 0) {
```

Comment:
nit: should we assert that `kEntropyCsrngBitsBufferSize` is a power of two and use `& (kEntropyCsrngBitsBufferSize - 1)` here?

Reply:
Done.